File: client/src/main/java/au/com/codeka/warworlds/client/game/fleets/ActionBottomPane.java
package au.com.codeka.warworlds.client.game.fleets;

import android.content.Context;
import android.graphics.Color;
import android.view.Gravity;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AbsListView;
import android.widget.BaseAdapter;
import android.widget.FrameLayout;
import android.widget.RelativeLayout;
import android.widget.Spinner;
import android.widget.SpinnerAdapter;
import android.widget.TextView;

import com.google.common.base.CaseFormat;

import javax.annotation.Nonnull;

import au.com.codeka.warworlds.client.R;
import au.com.codeka.warworlds.common.proto.Fleet;

import static com.google.common.base.Preconditions.checkNotNull;

/**
 * Bottom pane of the fleets view that contains the main action buttons (split, move, etc).
 */
public class ActionBottomPane extends RelativeLayout {
  /** Implement this to get notified of events. */
  public interface Callback {
    void onMoveClick();
    void onSplitClick();
  }

  private final Callback callback;

  public ActionBottomPane(Context context, @Nonnull Callback callback) {
    super(context, null);
    this.callback = checkNotNull(callback);

    inflate(context, R.layout.ctrl_fleet_action_bottom_pane, this);

    Spinner stanceSpinner = (Spinner) findViewById(R.id.stance);
    stanceSpinner.setAdapter(new StanceAdapter());

    findViewById(R.id.split_btn).setOnClickListener(this::onSplitClick);
    findViewById(R.id.move_btn).setOnClickListener(this::onMoveClick);
  }

  /** Called when you click 'split'. */
  private void onSplitClick(View view) {
    callback.onSplitClick();
  }

  /** Called when you click 'move'. */
  private void onMoveClick(View view) {
    callback.onMoveClick();
  }

  public class StanceAdapter extends BaseAdapter implements SpinnerAdapter {
    Fleet.FLEET_STANCE[] values;

    StanceAdapter() {
      values = Fleet.FLEET_STANCE.values();
    }

    @Override
    public int getCount() {
      return values.length;
    }

    @Override
    public Object getItem(int position) {
      return values[position];
    }

    @Override
    public long getItemId(int position) {
      return values[position].getValue();
    }

    @Override
    public View getView(int position, View convertView, ViewGroup parent) {
      TextView view = getCommonView(position, convertView, parent);

      view.setTextColor(Color.WHITE);
      return view;
    }

    @Override
    public View getDropDownView(int position, View convertView, ViewGroup parent) {
      TextView view = getCommonView(position, convertView, parent);

      ViewGroup.LayoutParams lp = new AbsListView.LayoutParams(FrameLayout.LayoutParams.MATCH_PARENT,
          FrameLayout.LayoutParams.MATCH_PARENT);
      lp.height = 80;
      view.setLayoutParams(lp);
      view.setTextColor(Color.WHITE);
      view.setText("  "+view.getText().toString());
      return view;
    }

    private TextView getCommonView(int position, View convertView, ViewGroup parent) {
      TextView view;
      if (convertView != null) {
        view = (TextView) convertView;
      } else {
        view = new TextView(getContext());
        view.setGravity(Gravity.CENTER_VERTICAL);
      }

      Fleet.FLEET_STANCE value = values[position];
      view.setText(CaseFormat.UPPER_UNDERSCORE.to(CaseFormat.UPPER_CAMEL, value.toString()));
      return view;
    }
  }
}

File: client/src/main/kotlin/au/com/codeka/warworlds/client/game/fleets/ActionBottomPane.kt
package au.com.codeka.warworlds.client.game.fleets

import android.content.Context
import android.graphics.Color
import android.view.Gravity
import android.view.View
import android.view.ViewGroup
import android.widget.*
import au.com.codeka.warworlds.client.R
import au.com.codeka.warworlds.common.proto.Fleet.FLEET_STANCE
import com.google.common.base.CaseFormat

/**
 * Bottom pane of the fleets view that contains the main action buttons (split, move, etc).
 */
class ActionBottomPane(context: Context, private val callback: Callback) : RelativeLayout(context, null) {
  /** Implement this to get notified of events.  */
  interface Callback {
    fun onMoveClick()
    fun onSplitClick()
    fun onMergeClick()
  }

  /** Called when you click 'split'.  */
  private fun onSplitClick() {
    callback.onSplitClick()
  }

  /** Called when you click 'merge'.  */
  private fun onMergeClick() {
    callback.onMergeClick()
  }

  /** Called when you click 'move'.  */
  private fun onMoveClick() {
    callback.onMoveClick()
  }

  inner class StanceAdapter internal constructor() : BaseAdapter(), SpinnerAdapter {
    var values: Array<FLEET_STANCE> = FLEET_STANCE.values()

    override fun getCount(): Int {
      return values.size
    }

    override fun getItem(position: Int): Any {
      return values[position]
    }

    override fun getItemId(position: Int): Long {
      return values[position].value.toLong()
    }

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
      val view = getCommonView(position, convertView)
      view.setTextColor(Color.WHITE)
      return view
    }

    override fun getDropDownView(position: Int, convertView: View?, parent: ViewGroup): View {
      val view = getCommonView(position, convertView)
      val lp: ViewGroup.LayoutParams =
          AbsListView.LayoutParams(
              FrameLayout.LayoutParams.MATCH_PARENT, FrameLayout.LayoutParams.MATCH_PARENT)
      lp.height = 80
      view.layoutParams = lp
      view.setTextColor(Color.WHITE)
      view.text = String.format("  %s", view.text.toString())
      return view
    }

    private fun getCommonView(position: Int, convertView: View?): TextView {
      val view: TextView
      if (convertView != null) {
        view = convertView as TextView
      } else {
        view = TextView(context)
        view.gravity = Gravity.CENTER_VERTICAL
      }
      val value = values[position]
      view.text = CaseFormat.UPPER_UNDERSCORE.to(CaseFormat.UPPER_CAMEL, value.toString())
      return view
    }

  }

  init {
    View.inflate(context, R.layout.ctrl_fleet_action_bottom_pane, this)
    val stanceSpinner = findViewById<View>(R.id.stance) as Spinner
    stanceSpinner.adapter = StanceAdapter()
    findViewById<View>(R.id.split_btn).setOnClickListener { onSplitClick() }
    findViewById<View>(R.id.merge_btn).setOnClickListener { onMergeClick() }
    findViewById<View>(R.id.move_btn).setOnClickListener { onMoveClick() }
  }
}